docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md

```diff
@@ -0,0 +1,198 @@
+### Existing PR
+[feat(compiler): allow explicit lift qualifications of preflight objects](https://github.com/winglang/wing/pull/5935)
+
+Issues with this PR:
+* Confusing sytax that looks like an inflight function call might lead to [misuse](https://github.com/winglang/wing/pull/5935#issuecomment-1995051115).
+* No scoping of where we're disabling compiler errors auto qualification.
+* No mechanism or clear path to mechanism for programatically defining and reusing lift qualifications.
+
+## Proposal
+
+```js
+let b1 = new cloud.Bucket() as "b1";
+let b2 = new cloud.Bucket() as "b2";
+inflight () => {
+  let var b = b1;
+  if true {
+    b = b2;
+  }
+  // In the following scope we disable qualification compilation errors due to unknown preflight object
+  // and we also disable automatic qualification.
+  // We explicitly qualify b1 and b2 with "put" since the compiler doesn't figure this out for us.
+  with lift_qualification {b1: ["put"], b2: ["put"]} {
+    b.put("key", "value"); 
+  }
+}
+```
+
+Internally the argument to the `lift_qualification` keyword creates a `std.Qualifications` object wich is added to the closures lift qualifications. 
+
+Note that multiple `lift_qualification` blocks in the code just create more `std.Qualifications`s for this closure. 
+The actuall qualification values (`b1` and `"put"`) are just appended to the closure/method and they don't strictly belong to the block.
+
+What does apply to the block are the disabling of qualification errors within the block, so in our example we can use the inflight expression `b`, and inside the block we also don't automatically add qualifications even if the compile can technically do this. 
+This behavior ensures that we can create custom qualifications without worrying that the compiler with override this with something more general.
+
+## Next steps
+
+The next step will be to provide use library APIs for programatically creating and manipulating `std.Qualifications` objects and then reusing them. We can also inherit from them and to provide qualification objects that go beyond only the "operation" being performed on the object. Here's an example:
+
+```js
+// stdlib
+class cloud.BucketPutQual extends std.Qualification {
+  this.keyPattern: str;
+  
+  new(b: cloud.Bucket) {
+    super(b, ["put"]);
+    this.keyPattern = "*";
+  }
+  
+  setKeyPattern(pattern: str) {
+    this.keyPattern = pattern;
+  }
+}
+
+// Create a qualifications container
+let quals = new std.Qualifications();
+
+
+let b1 = new cloud.Bucket() as "b1";
+let b2 = new cloud.Bucket() as "b1";
+
+// Define our explicit qualifications, b1 has a special key pattern when putting
+let b1Quals = new cloud.BucketPutCap(b1);
+b1Quals.setKeyPattern("k*");
+// b2 can put any key
+let b2Quals = new std.Qualification(b2, ["put"]);
+
+// Add them to our explicit qualifications
+quals.add(b1Quals);
+quals.add(b2Quals);
+
+let arr = [b1,b2];
```

Comment:
I think this should unblock our collection use cases:

```js
// creating a capability for invoking "put" on every bucket in an array
let buckets: Array<cloud.Bucket> = ...
let bucketsCap = new std.Capability();
for b in buckets {
  bucketsCap.add(b, ["put"]);
}

// later

inflight () => {
  with capability bucketsCap {
    for b in buckets {
      b.put("key", "value");
    }
  }
}
```

Reply:
right, added to the example